### PR TITLE
Path Tracer

### PR DESCRIPTION
**Pull request by Negrini085**
Here we want to actually write the path tracing algorithm. In order to do so, we have to implement Duff & All algorithm and to test it by means of random testing.

- [x] Ortho Normal Basis
- [x] Duff & All algorithm
- [x] Random testing

We than have to implement a scatter ray procedure for every kind of BRDF we have defined.

- [x] scatter ray Diffusive
- [x] scatter ray Specular 

Finally we have to implement path tracing procedure.

- [x] Path tracer

We also want to add different renderer kinds, such as:

- [x] OnOffRenderer
- [x] FlarRenderer

### CODE REVIEW

**Comment by Negrini085** [comment]
Flat renderer
Flat

**Comment by Negrini085** [comment]
Path tracer: i feel like something wrong with specularBRDF

Path

File: PhotoNim.nim [hunk 1]
```diff
@@ -126,36 +126,93 @@ let demoDoc* = """
 PhotoNim CLI `demo` command:
 
 Usage:
-    ./PhotoNim demo (persp | ortho) [<output>] [--w=<width> --h=<height> --angle=<angle>]
+    ./PhotoNim demo (persp | ortho) (OnOff | Flat | Path) [<output>] [--w=<width> --h=<height> --angle=<angle>]
```
Comment (lorenzoliuzzo) [inline PhotoNim.nim:129]: I would add a flag --renderer or something like that.

File: PhotoNim.nim [hunk 2]
```diff
@@ -126,36 +126,93 @@ let demoDoc* = """
...
 
     var 
-        scenary = World(shapes: @[s0, s1, s2, s3, s4, s5, s6, s7, s8, s9])
+        scenery = World(shapes: @[s0, s1, s2, s3, s4, s5, s6, s7, s8, s9])
         tracer = newImageTracer(width, height, camera, sideSamples = 4)
-
-    tracer.fire_all_rays(scenary, proc(ray: Ray): Color = newColor(0.3, 1.0, 0.2))
+    
+    render.world = scenery
+    tracer.fire_all_rays(render)
```
Comment (lorenzoliuzzo) [inline PhotoNim.nim:215]: ImageTracer should contain a renderer as member.

File: src/camera.nim [hunk 3]
```diff
@@ -72,8 +75,8 @@ type Ray* = object
     tmax*: float32
     depth*: int
 
-proc newRay*(origin: Point3D, direction: Vec3f): Ray {.inline.} = 
-    Ray(origin: origin, dir: direction, tmin: epsilon(float32), tmax: Inf, depth: 0)  
+proc newRay*(origin: Point3D, direction: Vec3f, tmin: float32 =  epsilon(float32), tmax = Inf, depth: int = 0): Ray {.inline.} = 
+    Ray(origin: origin, dir: direction, tmin: tmin, tmax: tmax, depth: depth)
```
Comment (lorenzoliuzzo) [inline src/camera.nim:79]: Use Ray() constructor instead of another newRay proc.

File: src/geometry.nim [hunk 4]
```diff
@@ -589,6 +590,29 @@ proc `/`*(transf: Transformation, scal: float32): Transformation =
...
+#-------------------------------------#
+
+type Onb* = object
+    base*: array[3, Vec3f]
+
+proc newONB*(e1: Vec3f = eX, e2: Vec3f = eY, e3: Vec3f = eZ): Onb {.inline.} =
+    Onb(base: [e1, e2, e3])
+
+
+proc create_onb*(norm: Normal): Onb =
```
Comment (lorenzoliuzzo) [inline src/geometry.nim:604]: I would call it newONB

File: src/geometry.nim [hunk 5]
```diff
@@ -589,6 +590,29 @@ proc `/`*(transf: Transformation, scal: float32): Transformation =
         return Transformation(kind: tkComposition, transformations: transfs)
 
 
+#-------------------------------------#
+#          OrthoNormal basis          #
+#-------------------------------------#
+
+type Onb* = object
```
Comment (lorenzoliuzzo) [inline src/geometry.nim:597]: Add a procedure to project a Vec in the ONB and viceversa.

File: src/tracer.nim [hunk 6]
```diff
@@ -60,4 +66,136 @@ proc fire_all_rays*(tracer: var ImageTracer; scenary: World, color_map: proc(ray
...
+proc newPathTracer*(world = newWorld(), back_col = newColor(0,0,0), randgen = newPCG(), n_ray = 10, max_depth = 10, roulette_lim = 3): Renderer {.inline.} =
+    Renderer(kind: PathTracer, world: world, back_col: back_col,
+            randgen: randgen, num_ray: n_ray, max_depth: max_depth, 
+            roulette_lim: roulette_lim) 
+
+
+#------------------------------------------------------------#
+#     Call procedure --> proc to solve rendering equation    #
+#------------------------------------------------------------#
+proc call*(rend: var Renderer, ray: Ray): Color =
```
Comment (lorenzoliuzzo) [inline src/tracer.nim:108]: I would call it `render`.

File: src/tracer.nim [hunk 7]
```diff
@@ -60,4 +66,136 @@ proc fire_all_rays*(tracer: var ImageTracer; scenary: World, color_map: proc(ray
...
+
+                for u in 0..<tracer.sideSamples:
+                    for v in 0..<tracer.sideSamples:
+                        let 
+                            pixelOffset = newPoint2D(
+                                (u.float32 + tracer.rg.rand) / tracer.sideSamples.float32, 
+                                (v.float32 + tracer.rg.rand) / tracer.sideSamples.float32
+                            )
+                                                
+                        color = color + rend.call(tracer.fire_ray(x, y, pixelOffset))
```
Comment (lorenzoliuzzo) [inline src/tracer.nim:192]: Use `+=` proc for Color.

**Review by lorenzoliuzzo — APPROVED** [review]

**Pull request merged**